### PR TITLE
Add support for Renzo Protocol

### PR DESCRIPTION
**Pull request by sunosuporno**
# Background
The Renzo plugin is being developed as part of the GOAT SDK to provide integration with Renzo Protocol, a restaking protocol.

## What does this PR do?
This PR implements a Renzo plugin that enables interaction with the Renzo protocol across multiple EVM chains (Mode, Base, Arbitrum, BSC, and Linea). The plugin provides three main functionalities:
1. Deposit/restake ERC20 LST tokens into Renzo
2. Deposit/restake native ETH into Renzo
3. Check ezETH balance for any address

## What kind of change is this?
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Documentation update

# Documentation changes needed?
My changes do not require a change to the project documentation.

# Testing
Can be done in `typescript/examples/vercel-ai/mode` file. Follow the docs to setup and then enter prompts to test out the plugin

## Detailed testing results
| Function | Prompt | Proof | Transaction Link |
|----------|--------|-------|------------------|
| deposit LSTs | can you deposit 0.001 WETH erc20 token in Renzo? |Screenshot 2024-12-21 at 9 23 34 PM | [View txn](https://explorer.mode.network/tx/0xbf9783c03fe9885ada45950eae0a072d2a6eb93a89ab004af478250ee2f40c53)|
| depositETH | can you send 0.001 ETH to deposit it in Renzo? | Screenshot 2024-12-21 at 10 05 28 PM | [View txn](https://explorer.mode.network/tx/0x8e8cd6287c10c8a72f81379e4f97e335437686df3d14a7703b767b6f7bdff1b0)|
| ezETH balanceOf | can you check the ezETH balance for my wallet address in Renzo? | Screenshot 2024-12-21 at 9 54 58 PM | |

## For plugins
- [x] I have tested this change locally with key pair wallets
- [ ] I have tested this change locally with hosted wallets (e.g. Crossmint Smart Wallets, etc.)

## Discord username
sunosuporno

### CODE REVIEW

File: typescript/packages/plugins/renzo/src/renzo.service.ts [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+        name: "renzo_deposit_erc20",
+        description:
+            "Deposit ERC20 LST tokens into Renzo, approve the ERC20 contract to spend the tokens before calling this",
+    })
+    async depositERC20(
+        walletClient: EVMWalletClient,
+        parameters: DepositParams
+    ) {
+        try {
+            console.log(
```
Comment (0xaguspunk) [inline typescript/packages/plugins/renzo/src/renzo.service.ts:27]: No `console.logs` in the plugins! Good for debugging but no need for that on the agent :)

File: typescript/packages/plugins/renzo/src/parameters.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
+import { createToolParameters } from "@goat-sdk/core";
+import { z } from "zod";
+
+export class DepositParams extends createToolParameters(
+    z.object({
+        _token: z.string().describe("The token to deposit"),
```
Comment (0xaguspunk) [inline typescript/packages/plugins/renzo/src/parameters.ts:6]: Do parameters need the underscore?

File: typescript/packages/plugins/renzo/src/parameters.ts [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+        _amountIn: z.string().describe("The amount of tokens to deposit"),
```
Comment (0xaguspunk) [inline typescript/packages/plugins/renzo/src/parameters.ts:7]: I would add "in basis units"

File: typescript/packages/plugins/renzo/src/parameters.ts [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+        _minOut: z.string().describe("The minimum amount of tokens to receive"),
```
Comment (0xaguspunk) [inline typescript/packages/plugins/renzo/src/parameters.ts:8]: Same here!

File: typescript/packages/plugins/renzo/src/parameters.ts [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+        _deadline: z.string().describe("The deadline for the deposit"),
```
Comment (0xaguspunk) [inline typescript/packages/plugins/renzo/src/parameters.ts:9]: Which format does this follow? Check how we did it here in the KIM plugin: 
* https://github.com/goat-sdk/goat/blob/main/typescript/packages/plugins/kim/src/parameters.ts#L38
* https://github.com/goat-sdk/goat/blob/main/typescript/packages/plugins/kim/src/kim.service.ts#L48

File: typescript/packages/plugins/renzo/src/renzo.service.ts [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+
+            return hash.hash;
+        } catch (error) {
+            console.error(`[Renzo] Deposit failed with error:`, error);
+            throw Error(`Failed to deposit ERC20: ${error}`);
+        }
+    }
+
+    @Tool({
+        name: "renzo_deposit_eth",
```
Comment (0xaguspunk) [inline typescript/packages/plugins/renzo/src/renzo.service.ts:109]: Maybe the name can be imperative like the description. `deposit_eth_in_renzo`

File: typescript/packages/plugins/renzo/src/renzo.service.ts [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+
+            const { renzoDepositAddress } = getRenzoAddresses(
+                walletClient.getChain().id
+            );
+            console.log(
+                `[Renzo] Using deposit address: ${renzoDepositAddress}`
+            );
+
+            const _minOut = parseUnits(parameters._minOut, 18);
+            const value = parseUnits(parameters._value, 18);
```
Comment (0xaguspunk) [inline typescript/packages/plugins/renzo/src/renzo.service.ts:130]: No need to add this. You can receive it in basis units and there is a tool that already converts into basis units in the erc20 plugin

File: typescript/packages/plugins/renzo/src/renzo.service.ts [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+
+            return hash.hash;
+        } catch (error) {
+            console.error(`[Renzo] ETH deposit failed with error:`, error);
+            throw Error(`Failed to deposit ETH: ${error}`);
+        }
+    }
+
+    @Tool({
+        name: "renzo_check_ezeth_balance",
```
Comment (0xaguspunk) [inline typescript/packages/plugins/renzo/src/renzo.service.ts:159]: Same here with the name :)

File: typescript/packages/plugins/renzo/src/renzo.service.ts [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+            console.log(`[Renzo] Balance result type:`, typeof balanceResult);
+            console.log(
+                `[Renzo] Balance result value:`,
+                JSON.stringify(balanceResult, (_, value) =>
+                    typeof value === "bigint" ? value.toString() : value
+                )
+            );
+
+            const balance = (balanceResult as { value: bigint }).value;
+            const formattedBalance = formatUnits(balance, 18);
```
Comment (0xaguspunk) [inline typescript/packages/plugins/renzo/src/renzo.service.ts:195]: No need to do this either there is a tool that converts it back to decimals

File: typescript/packages/plugins/renzo/src/renzo.service.ts [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+        } catch (error) {
+            console.error(`[Renzo] Failed to get ezETH balance:`, error);
+            throw Error(`Failed to get ezETH balance: ${error}`);
+        }
+    }
+
+    @Tool({
+        name: "renzo_approve_deposit",
+        description:
+            "Approve Renzo deposit contract to spend tokens. Call this before calling erc20 deposit function.",
```
Comment (0xaguspunk) [inline typescript/packages/plugins/renzo/src/renzo.service.ts:210]: No need for this function, erc20 already approves. It just needs to have the token into

**Review by 0xaguspunk — CHANGES_REQUESTED** [review]
Amazing PR! Love the example tx and screenshots, will add it to the PR template :)

Made some comments, overall:
1. No console.logs/errors
2. No need for converting units into basis units, this is already handled by the erc20 plugin tool convert units
3. No need for the approve function, also handled by the erc20 plugin
4. I would add the token info to the `tokens.ts` file inside the erc20 plugin :)

File: typescript/packages/plugins/renzo/src/parameters.ts (hunk 5 above)
Comment (sunosuporno) [reply typescript/packages/plugins/renzo/src/parameters.ts:9]: Implemented the first one. Much easier I felt

**Comment by sunosuporno** [comment]
@0xaguspunk implemented the changes! The `depositETH` is not functioning when I remove the `parseUnits` calculations. I guess it is because ETH is not an ERC20. Rest everything is done I hope 😅

File: typescript/packages/plugins/renzo/package.json [hunk 11]
```diff
@@ -0,0 +1,42 @@
+{
+    "name": "@goat-sdk/plugin-renzo",
+    "version": "1.0.0",
```
Comment (0xaguspunk) [inline typescript/packages/plugins/renzo/package.json:3]: Version should be 0.1.0

File: typescript/packages/plugins/renzo/src/abi/erc20.ts
Comment (0xaguspunk) [inline typescript/packages/plugins/renzo/src/abi/erc20.ts:1]: This file is not being used so no need for it :)

**Review by 0xaguspunk — CHANGES_REQUESTED** [review]

**Comment by 0xaguspunk** [comment]
Made two small comments.

Also you need to delete the pnpm-lock file and run `pnpm install` again from the `typescript` directory!

Once done we can merge! :)

**Comment by sunosuporno** [comment]
@0xaguspunk Made the changes!

**Review by 0xaguspunk — APPROVED** [review]

**Comment by fede2442** [comment]
Hey @sunosuporno congrats on getting this PR merged!

Please share an EVM wallet to receive the payment for the [Mode-GOAT Bounties](https://modenetwork.notion.site/GOAT-Plugins-Bounties-15ce5fc9bc99805aa41ec75876fa65ba?pvs=4)

**Comment by sunosuporno** [comment]
Hello @fede2442, my wallet address is 0x07a721260416e764618B059811eaf099a940Af14 🙌

**Comment by fede2442** [comment]
Payment has been sent along with the other plugin bounties you did, here is the tx hash:

https://explorer.mode.network/tx/0x5c7e97d6826806df71b1c35bfdbd34b2e538b02c21d2f5a0378a4f99a328f93f

**Pull request merged**